frontend/src/metabase/components/Modal.jsx

```diff
@@ -84,91 +75,91 @@ export class WindowModal extends Component {
-            key="modal"
-            className={cx(backdropClassName, backdropClassnames)}
-            style={style}
-          >
-            {this._modalComponent()}
-          </div>
-        )}
-      </CSSTransitionGroup>,
-      this._modalElement,
-    );
-  }
 
-  render() {
-    return null;
+    return (
+      <SandboxedPortal container={this._modalElement}>
+        <CSSTransitionGroup
+          transitionName="Modal"
+          transitionAppear={true}
+          transitionAppearTimeout={250}
+          transitionEnterTimeout={250}
+          transitionLeaveTimeout={250}
+        >
+          {isOpen && (
+            <div
+              key="modal"
+              className={cx(backdropClassName, backdropClassnames)}
+              style={style}
+            >
+              {this._modalComponent()}
+            </div>
+          )}
+        </CSSTransitionGroup>
+      </SandboxedPortal>
+    );
   }
 }
 
 import routeless from "metabase/hoc/Routeless";
 
 export class FullPageModal extends Component {
-  componentDidMount() {
+  constructor(props) {
+    super(props);
+    this.state = {
+      isOpen: true,
+    };
+
     this._modalElement = document.createElement("div");
     this._modalElement.className = "ModalContainer";
-    document.querySelector("body").appendChild(this._modalElement);
+    document.body.appendChild(this._modalElement);
 
     // save the scroll position, scroll to the top left, and disable scrolling
     this._scrollX = getScrollX();
     this._scrollY = getScrollY();
     window.scrollTo(0, 0);
     document.body.style.overflow = "hidden";
-
-    this.componentDidUpdate();
   }
 
-  componentDidUpdate() {
-    // set the top of the modal to the bottom of the nav
+  setTopOfModalToBottomOfNav() {
     const nav = document.body.querySelector(".Nav");
     if (nav) {
       this._modalElement.style.top = nav.getBoundingClientRect().bottom + "px";
     }
-    this._renderModal(true);
+  }
+
+  componentDidMount() {
+    this.setTopOfModalToBottomOfNav();
+  }
+
+  componentDidUpdate() {
+    this.setTopOfModalToBottomOfNav();
   }
 
   componentWillUnmount() {
-    this._renderModal(false);
+    this._modalElement.parentNode.removeChild(this._modalElement);
+  }
+
+  handleDismissal = () => {
+    this.setState({ isOpen: false });
 
     // restore scroll position and scrolling
     document.body.style.overflow = "";
-
     // On IE11 a timeout is required for the scroll to happen after the change of overflow setting
     setTimeout(() => {
       window.scrollTo(this._scrollX, this._scrollY);
     }, 0);
 
     // wait for animations to complete before unmounting
-    setTimeout(() => {
-      ReactDOM.unmountComponentAtNode(this._modalElement);
-      this._modalElement.parentNode.removeChild(this._modalElement);
-    }, 300);
-  }
-
-  handleDismissal = () => {
-    if (this.props.onClose) {
-      this.props.onClose();
-    }
+    setTimeout(() => this.props.onClose && this.props.onClose(), 300);
```

Comment:
Does `onClose` need the 300ms delay? I guess the unmounting for `FullPageModal` actually happens in `Modal` so maybe it should live there to delay removing it from the DOM. Though, if there are no issues with exit animations, maybe it's not needed?

Reply:
`FullPageModal` / `Modal` are unmounted from the parent component, whatever is passing the `onClose` prop. Without the delay the close animation doesn't work when the use presses the "esc" key. The animation when the user presses the X icon is (and has been) broken because the parent component passes the `onClose` prop directly to the child component of `FullScreenModal`, so we have no way of delaying things.

I'll create an issue to revisit the missing `WindowModal` entry animation. Besides that, these modals could definitely use some work to improve consistency.